### ISSUE TITLE
Benjamin Fleischer Extension Won't Install on M1 MacBook Pro

### ISSUE THREAD

**Issue by ExactObject** [osxfuse/osxfuse#754]
I'm running the 11.1 Beta (20C5048k) on a M1 Macbook Pro.  When installing MacFuse 4.0.4, the Benjamin Fleischer extension is blocked and so use the Security & Privacy preference to allow it.  After allowing, the OS wants a reboot to rebuild the extension library, and about 80% through the reboot process, a silent failure occurs and the system reboots by itself.  Head back to the Security & Privacy preference pane and it's still asking to allow the Benjamin Fleischer extension.  Rinse & repeat; then uninstall MacFuse to get it to stop asking to allow the Benjamin Fleischer extension.

**Comment by djkcel** [comment]
I'm running the same os on an M1 MacBook Air.

I had to boot into recovery mode and then launch the "Startup Security Utility." From here I enabled kernel extensions and rebooted, after which I was able to "allow" in the security prefpane. However it still doesn't work. When I try to actually mount a volume in terminal I get 

`"kext load failed: -603947007"`

After trying to manually load the kext, I get:

```
Executing: /usr/bin/kmutil load --bundle-path /Library/Filesystems/osxfuse.fs/Contents/Extensions/10.11/osxfuse.kext
Error Domain=KMErrorDomain Code=71 "Incompatible architecture: Binary is for x86_64, but needed arch arm64e" UserInfo={NSLocalizedDescription=Incompatible architecture: Binary is for x86_64, but needed arch arm64e}
```

I fear that this software may not be compatible with the new Apple M1 ARM chips but I'll look around a little more.

**Comment by vijaynr** [comment]
@djkcel  are you sure you are loading the correct kext? I can see **osxfuse is renamed to macfuse** in versions 4+. 
The kext path would be something like this `/Library/Filesystems/macfuse.fs/Contents/Extensions/11/macfuse.kext`

The one you are trying to load might have **Intel only support**. Please check.

**Comment by bfleischer** [comment]
@djkcel The version of FUSE you are using is not compatible with Apple Silicon Macs. Please upgrade to macFUSE 4

**Comment by tommythorn** [comment]
UPDATE 2: I don't know what I did wrong but`kmutil unload --bundle-path /Library/Filesystems/macfuse.fs/Contents/Extensions/11/macfuse.kext` _does_ work and will carry me over until the next version (now to see if I can make Keybase work with it).

**Comment by NewAlexandria** [comment]
I have this issue, but with version 4.0.4.  I am testing on a new MacBookPro (silicon / M1), Big Sur (obviously).  
* I tried installing normally + security panel to accept.   (not-working)
* I also tried `csrutil disable` after an uninstall of macfuse, then reinstalling, security panel allow, and related requisite reboots (not-working)
* I tried manually loading the kext via the terminal.  This worked to stop the security panel alerts (the 'allow app'), but my normal client apps (Keybase) still won't run.  

How can I confirm if MacFUSE is working correctly?  I don't normally use this library otherwise, and there's no 'hello world' that I see in docs.  Thanks

**Comment by veganmaister** [comment]
> @djkcel The version of FUSE you are using is not compatible with Apple Silicon Macs. Please upgrade to macFUSE 4

I’m experiencing the same problem with 4.0.4 which has subsequently broken pCloud.

**Comment by veganmaister** [comment]
Is there any way we can help fast track the resolution here? I find it astounding how a little OS extension has completely broken productivity on M1 Macs.

**Comment by bfleischer** [comment]
macFUSE 4 is required on M1 Macs. Older releases of FUSE do not support the new processor architecture. Please keep in mind that Rosetta, which translates old Intel apps to run on M1 Macs, does not support translating kernel extensions. That's why you will need to upgrade to macFUSE 4. 

@veganmaister There is no need downvote this. It's a matter of fact.

In macOS 11 Apple completely rewrote the Kext Manager component of macOS. Unfortunately, with all new software there are bugs. Since the first release of macOS 11, I raised 11 issues with Apple. Four of the issues have been addressed in the meantime. Seven are currently open. For three of these seven remaining issues, FUSE contains workarounds. This leaves 4 issues that can prevent third-party kernel extension from being loaded. This is true for macFUSE and pCloud.

**Comment by tommythorn** [comment]
@bfleischer Thanks for the informative update.  This is the kind of things we should expect with the M1 transition and I'm glad progress is being made.  Is subscribing to this bug the best way to monitor the progress?

**Comment by someone857** [comment]
I'm having the same or similar issue. I get to this loop where System Preferences wants me to "Allow" every time, and after a restart there's a popup stating "System Extension Error" and telling me it needs to rebuild, and then there's the Allow button again.
Not sure if that makes a difference, but I partitioned my macOS with 2 Volumes. In the first (original) I was able to install macFUSE 4.0.5 without a hitch (I think it was even before I created the 2nd volume, not sure atm). Trying to install macFUSE on my 2nd volume is where I'm getting this behaviour. I've authorized user extensions on both Volumes using the Startup Recovery.

**Comment by bfleischer** [comment]
@someone857 This is a known bug in macOS. It's unclear what triggers it. Apple is working on a fix. As a workaround you can try triggering the new panic medic.

To do so, run the following command in Terminal in the recovery environment. You will probably need to update the root volume path (/Volumes/System) since you have two macOS installs.

```
kmutil --trigger-panic-medic --volume-root /Volumes/System
```

This disables loading third-party kernel extensions until the user approves them again. This has worked for me in one or two cases when the Kext Manager was stuck.

**Comment by someone857** [comment]
Thanks for your reply @bfleischer . I saw your comment on https://github.com/osxfuse/osxfuse/issues/779#issuecomment-769719520 shortly after I wrote this. I tried running this code, but it gave me the same results (and btw trigger-panic-medic is not a flag, so the actual command is `kmutil trigger-panic-medic --volume-root /Volumes/MyVolume`).
When I logged to my OS it told me the extensions were reset, that every permission would be asked again, but once I tried to re-run macFUSE it gave me the same loop. Tried possibly every combination of installing/uninstalling and running this command and never got it working :(

From your comment, I think my only option is to wait for an Apple fix?

**Comment by bfleischer** [comment]
@someone857 Yes, the only proper solution is for Apple to address the bug. In the meantime you can try uninstalling macFUSE, triggering the panic medic, rebooting a couple of times and then reinstalling macFUSE. At some point macOS will load the extension.

**Comment by crmscenarios** [comment]
This thread on Apple's boards helped me out:
https://discussions.apple.com/thread/252261556

**Comment by mrjayberry** [comment]
@someone857 did you ever find a solution to getting macFUSE working on your 2nd partition? 
Extension won't install running 11.4 Big Sur on an external SSD, and the question of approving it from Security & Privacy just never appears.  
Yet it installs (and asks for approval) just fine on the OS running from the internal SSD. Both instances are otherwise identical clean installs. 

I've tried quite everything I can think of, to load the extension manually or trigger it asking me to approve in Security & Privacy: (of course Reduced Security is setup in recovery mode first)
- install/uninstall with 5 reboots between each step. 
- turning off SIP (and repeating.. install/uninstall/rebooting many times)
- triggering panic medic (and repeating install/uninstall). It does ask me to authorize other extensions, but not Benjamin. 
- tried 4.0.4, 4.1.2, 4.2 versions.. all same. 

However something interesting,
`kmutil unload --bundle-path /Library/Filesystems/macfuse.fs/Contents/Extensions/11/macfuse.kext` 
gives me the error 
Error Domain=NSPOSIXErrorDomain Code=1 "Operation not permitted"

Any advice?  
Note, I have limited experience using terminal, coding, etc. Just perseverance. Maybe I'm making an easy mistake somewhere..

**Comment by danielslyman** [comment]
Is there any news on this?

**Comment by atYuguo** [comment]
> 
> 
> @someone857 did you ever find a solution to getting macFUSE working on your 2nd partition?
> Extension won't install running 11.4 Big Sur on an external SSD, and the question of approving it from Security & Privacy just never appears.
> Yet it installs (and asks for approval) just fine on the OS running from the internal SSD. Both instances are otherwise identical clean installs.
> 
> I've tried quite everything I can think of, to load the extension manually or trigger it asking me to approve in Security & Privacy: (of course Reduced Security is setup in recovery mode first)
> 
>     * install/uninstall with 5 reboots between each step.
> 
>     * turning off SIP (and repeating.. install/uninstall/rebooting many times)
> 
>     * triggering panic medic (and repeating install/uninstall). It does ask me to authorize other extensions, but not Benjamin.
> 
>     * tried 4.0.4, 4.1.2, 4.2 versions.. all same.
> 
> 
> However something interesting,
> `kmutil unload --bundle-path /Library/Filesystems/macfuse.fs/Contents/Extensions/11/macfuse.kext`
> gives me the error
> Error Domain=NSPOSIXErrorDomain Code=1 "Operation not permitted"
> 
> Any advice?
> Note, I have limited experience using terminal, coding, etc. Just perseverance. Maybe I'm making an easy mistake somewhere..

Just for the `kmutil` command, you might need `sudo` in front of it for permissions, that is:
```
sudo kmutil unload --bundle-path /Library/Filesystems/macfuse.fs/Contents/Extensions/11/macfuse.kext
```

**Comment by atYuguo** [comment]
> 
> 
> Is there any news on this?

To install kernel extensions on a M1 Macbook, you need to run the system under "reduced security", see:

https://support.apple.com/guide/deployment-reference-macos/startup-security-ior2b1833593/web

**Comment by mrjayberry** [comment]
Thanks for your suggestion, I think it would have helped. :). Eventually I
returned the M1 and upgraded for another with larger internal SSD. So I
didn’t need to run the OS on an external SSD anymore.  Problem not solved,
but avoided.

**Comment by mrjayberry** [comment]
Thanks for your suggestion, I think it would have helped. :). Eventually I returned the M1 and upgraded for another with larger internal SSD. So I didn’t need to run the OS on an external SSD anymore. Problem not solved, but avoided.

**Comment by viking1304** [comment]
On my Intel MacBook Pro both 4.1.2 and 4.2.0 worked without any problem, but when I got M1 last week I had to overcome few problems before it stated working.

The main problem is that disabling Gatekeeper and allowing extension is not enough for M1 Macs. You need to modify security settings in the Recovery Environment first (as atYuguo wrote). You can find more information how to do that on this [link](https://www.m3datarecovery.com/mac-bitlocker/enable-system-extension-m1-mac.html) (steps to enable macFUSE are basically the same as on that page).

There is one more important thing you should do. When you enable extension after modifying those settings, I advice to click on Restart button. This will immediately trigger the clear of extension cache.

After this macFUSE should work fine on your M1 Macs.

I guess it should be good to add this info on FAQ.

**Comment by zanjie1999** [comment]
if you want to install some "requires closed-source macFUSE" software, such as bindfs, use: 
```
nano `brew formula bindfs`
```
and find 
```
  on_macos do
    disable! date: "2021-04-08", because: "requires closed-source macFUSE"
  end
```
delete it and Ctrl+x to save file, then you can install

**Comment by bkmcfarland** [comment]
how can this have been closed when the problem still persists with macfuse 4.4 running on macOS Monterey 12.4?? Possibly someone could post the bug number(s) for the issues with kernel extension consent with macos ??

**Comment by skkii** [comment]
> On my Intel MacBook Pro both 4.1.2 and 4.2.0 worked without any problem, but when I got M1 last week I had to overcome few problems before it stated working.
> 
> The main problem is that disabling Gatekeeper and allowing extension is not enough for M1 Macs. You need to modify security settings in the Recovery Environment first (as atYuguo wrote). You can find more information how to do that on this [link](https://www.m3datarecovery.com/mac-bitlocker/enable-system-extension-m1-mac.html) (steps to enable macFUSE are basically the same as on that page).
> 
> There is one more important thing you should do. When you enable extension after modifying those settings, I advice to click on Restart button. This will immediately trigger the clear of extension cache.
> 
> After this macFUSE should work fine on your M1 Macs.
> 
> I guess it should be good to add this info on FAQ.

Sorry to say, but this solution only works for a Part of the users. Many of us work primarily on business/corporate hardware. Here the option to disable Gatekeeper and/or security settings is simply not possible.

**Comment by tjluoma** [comment]
Nobody should be disabling Gatekeeper for any reason. Apps need to update or cease to exist if they can't function with Gatekeeper.

**Comment by bfleischer** [comment]
@tjluoma You don't need to disable Gatekeeper or System Integrity Protection to use macFUSE. I can't stress this enough.

----

**Intel**

On Intel Macs you get a prompt when trying to load the macFUSE kernel extension for the first time. You'll need to click on the 'Allow' button in the system preferences, then reboot your Mac and you are good to go.

**Apple Silicon**

On Apple Silicon Macs you get a prompt when trying to load the macFUSE kernel extension for the first time, just like on Intel Macs. Then you'll need to open system preferences and go to the Security and Privacy preference pane, just like on Intel Macs. 

By default, third-party kernel extensions are disabled on Apple Silicon Macs. If you have not already enabled support for third-party kernel extensions, follow the instructions displayed in the Security and Privacy preference pane. This involves rebooting your Mac into recovery mode.

When you are back in macOS, try mounting the macFUSE-based file system again. You will receive another prompt and need to click on the "Allow" in the system preferences, then reboot your Mac and you are good to go.

----

Always make sure to use the latest version of macFUSE and to install all available macOS updates.

If the above procedure does not work for you, you are most likely running into a macOS bug. Some users suggest disabling System Integrity Protection to get the kernel extension to load on affected Macs. While this seems to work reliably, it is not recommended. Disabling System Integrity protection bypasses the macOS bug, but it also reduces the security of your Mac greatly.

I recommend reporting the issue to Apple. You can use Apple's Feedback Assistent app to do so. Use Spotlight to open it or navigate to /System/Library/CoreServices/Applications in Finder.

**Comment by bfleischer** [comment]
Installation instructions for macFUSE with screenshots: https://github.com/macfuse/macfuse/wiki/Getting-Started

**Comment by Anatoly4444** [comment]
Can't give access for macfuse. Old way to make it on Ventura does not work

**Comment by schmodd** [comment]
@bfleischer, just for clarification: is it enough to enable third party kernel extensions, then allow macfuse in mac os and turn back to full security (and disable third party kernel extensions)? Or is it necessary to leave third party kernel extensions settings activated?

**Comment by bfleischer** [comment]
@schmodd You need to keep third-party kernel extensions enabled, otherwise macFUSE will stop working.

**Comment by kfattic** [comment]
I am running into this issue as well... I manage our MDM (Mosyle) at our EDU.  I have the Apple Silicon M1 in Reduced Security with both checkboxes checked.  I have the Developer ID: 3T5GSNBU6W allowed in both Kernel Extensions and System Extensions.  For the Kernel Extension profile I have both Allow User Override and Allow Non-Admin User Approvals checked as well.  This WAS working at some point in time on an earlier version of Monterey.  When this issue came up again, the machine had macOS Monterey 12.6.3 on it and wasn't working.  I then upgraded it to macOS 13.2.1 Ventura and still no dice after spending an hour or working on it in Monterey.  Computer has macfuse 4.4.2 on it now.  As an aside, the last time I got this to work, I actually had to exclude the machine from the Kernel Extension approval and was able to get the message in the Security & Privacy settings to appear and allow manual approval, but I am unable to get that to appear again.
